### PR TITLE
[JENKINS-38831] Add support for credential tracking in SecretBuildWrapper and BindingStep

### PR DESCRIPTION
**Pull request by iwarapter**
[JENKINS-38831](https://issues.jenkins-ci.org/browse/JENKINS-38831)

### CODE REVIEW

**Comment by iwarapter** [comment]
This is currently setup to track the instances of builds using the credential, would we want this to be that granular or do we just want the Job?

File: src/main/java/org/jenkinsci/plugins/credentialsbinding/MultiBinding.java [hunk 1]
```diff
@@ -62,7 +62,7 @@ protected MultiBinding(String credentialsId) {
     }
 
     /** Type token. */
-    protected abstract Class<C> type();
+    public abstract Class<C> type();
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/credentialsbinding/MultiBinding.java]: Binary-incompatible for external implementations, I think.

Probably solvable by adding a `public final` method to `MultiBinding` to set up the tracking.

File: src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java [hunk 2]
```diff
@@ -92,6 +94,8 @@
                 MultiBinding.MultiEnvironment environment = binding.bind(run, workspace, launcher, listener);
                 unbinders.add(environment.getUnbinder());
                 overrides.putAll(environment.getValues());
+                IdCredentials id = CredentialsProvider.findCredentialById(binding.getCredentialsId(), binding.type(), run);
+                CredentialsProvider.track(run, id);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java]: Perhaps also desirable to call `track(Node, Credentials)`. (For `BindingStep`, just inject a `Node`; for `SecretBuildWrapper`, use the method on `AbstractBuild` IIRC.) @stephenc any guidance?

Comment (stephenc) [inline src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java]: So track(Node, C) is to track when the credential is used *to launch that node*. Not the case here. We already have tracking between the Run and the Node(s) it ran on.

File: src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java [hunk 3]
```diff
@@ -92,6 +94,8 @@
                 MultiBinding.MultiEnvironment environment = binding.bind(run, workspace, launcher, listener);
                 unbinders.add(environment.getUnbinder());
                 overrides.putAll(environment.getValues());
+                IdCredentials id = CredentialsProvider.findCredentialById(binding.getCredentialsId(), binding.type(), run);
```
Comment (stephenc) [inline src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java]: Ugh! The tracking should be done when the credential is actually looked up... not by doing a second look up... as that second look-up may trigger a second logged call to an external credentials management system. Much better if we can have it as part of the binding... May require a thread local to capture the Run context

File: src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper.java [hunk 4]
```diff
@@ -57,6 +59,8 @@
         final List<MultiBinding.MultiEnvironment> m = new ArrayList<MultiBinding.MultiEnvironment>();
         for (MultiBinding binding : bindings) {
             m.add(binding.bind(build, build.getWorkspace(), launcher, listener));
+            IdCredentials id = CredentialsProvider.findCredentialById(binding.getCredentialsId(), binding.type(), build);
```
Comment (stephenc) [inline src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/SecretBuildWrapper.java]: Ugh! Please no... one lookup only

**Comment by iwarapter** [comment]
@jglick @stephenc thanks for your feedback, this should fix all the issues i believe.

**Comment by iwarapter** [comment]
The jenkins hook seems to be stuck, can someone give it a kick :)

**Comment by iwarapter** [comment]
Is there anything else needed for this PR?

File: src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java (hunk 2 above)
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java]: Well, not in the case of Pipeline builds generally, but I see your point—`track(Node, Credentials)` is irrelevant here. I will file a PR to fix the Javadoc.

File: src/main/java/org/jenkinsci/plugins/credentialsbinding/MultiBinding.java [hunk 5]
```diff
@@ -24,17 +24,13 @@
...
 import com.cloudbees.plugins.credentials.common.IdCredentials;
 import com.cloudbees.plugins.credentials.common.StandardCredentials;
 import hudson.ExtensionPoint;
 import hudson.FilePath;
 import hudson.Launcher;
-import hudson.model.AbstractDescribableImpl;
-import hudson.model.Descriptor;
-import hudson.model.Run;
-import hudson.model.TaskListener;
+import hudson.model.*;
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/credentialsbinding/MultiBinding.java]: Please revert and use explicit imports.

File: src/main/java/org/jenkinsci/plugins/credentialsbinding/MultiBinding.java [hunk 6]
```diff
@@ -120,6 +116,7 @@ public Unbinder getUnbinder() {
      */
     protected final @Nonnull C getCredentials(@Nonnull Run<?,?> build) throws IOException {
         IdCredentials cred = CredentialsProvider.findCredentialById(credentialsId, IdCredentials.class, build);
+        CredentialsProvider.track(build, cred);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/credentialsbinding/MultiBinding.java]: Better moved to the `if` block that `return`s a value.

File: src/test/java/org/jenkinsci/plugins/credentialsbinding/impl/UsernamePasswordBindingTest.java [hunk 7]
```diff
@@ -24,31 +24,39 @@
 
 package org.jenkinsci.plugins.credentialsbinding.impl;
 
-import com.cloudbees.plugins.credentials.CredentialsProvider;
-import com.cloudbees.plugins.credentials.CredentialsScope;
+import com.cloudbees.plugins.credentials.*;
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/credentialsbinding/impl/UsernamePasswordBindingTest.java]: Ditto, please retain explicit imports.

**Comment by iwarapter** [comment]
@jglick updated from feedback.

**Pull request merged**